File: 2 pointers/2486_append_characters_to_string_to_make_subsequence.py
import unittest

"""

You are given two strings s and t consisting of only lowercase English letters.

Return the minimum number of characters that need to be appended to the end of s so that t becomes a subsequence of s.

A subsequence is a string that can be derived from another string by deleting some or no characters without changing the order of the remaining characters.

"""

class Solution:
    def appendCharacters(self, s: str, t: str) -> int:

        N = len(s)
        M = len(t)
        point1, point2 = 0, 0

        while point1 < N and point2 < M:

            if s[point1] == t[point2]:
                point1 += 1
                point2 += 1
            else:
                point1 += 1

        return M - point2


class TestAppendCharacters(unittest.TestCase):
    def setUp(self):
        self.solution = Solution()

    def test_cases(self):
        # Test case 1: s and t are identical
        self.assertEqual(self.solution.appendCharacters("abc", "abc"), 0)

        # Test case 2: t is a substring of s
        self.assertEqual(self.solution.appendCharacters("abcdef", "ace"), 0)

        # Test case 3: t is not present in s at all
        self.assertEqual(self.solution.appendCharacters("abcdef", "xyz"), 3)

        # Test case 4: s is empty
        self.assertEqual(self.solution.appendCharacters("", "xyz"), 3)

        # Test case 5: t is empty
        self.assertEqual(self.solution.appendCharacters("abcdef", ""), 0)

        # Test case 6: s is shorter than t but has some matching characters
        self.assertEqual(self.solution.appendCharacters("ace", "abcdef"), 5)

        # Test case 7: s is much longer than t with t scattered within s
        self.assertEqual(self.solution.appendCharacters("abacadaeaf", "aaa"), 0)

        # Test case 8: All characters in t are in s but not in order
        self.assertEqual(self.solution.appendCharacters("abacadaeaf", "fa"), 1)

        # Test case 9: Partial match at the beginning of s
        self.assertEqual(self.solution.appendCharacters("abc", "abf"), 1)

        # Test case 10: Partial match at the end of s
        self.assertEqual(self.solution.appendCharacters("abcdef", "defg"), 1)


if __name__ == "__main__":
    unittest.main()

File: 2 pointers/2486_append_characters_to_string_to_make_subsequence.rs

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_identical_strings() {
        let s = String::from("abc");
        let t = String::from("abc");
        assert_eq!(Solution::append_characters(s, t), 0);
    }

    #[test]
    fn test_t_is_substring_of_s() {
        let s = String::from("abcdef");
        let t = String::from("ace");
        assert_eq!(Solution::append_characters(s, t), 0);
    }

    #[test]
    fn test_t_not_present_in_s() {
        let s = String::from("abcdef");
        let t = String::from("xyz");
        assert_eq!(Solution::append_characters(s, t), 3);
    }

    #[test]
    fn test_empty_s() {
        let s = String::from("");
        let t = String::from("xyz");
        assert_eq!(Solution::append_characters(s, t), 3);
    }

    #[test]
    fn test_empty_t() {
        let s = String::from("abcdef");
        let t = String::from("");
        assert_eq!(Solution::append_characters(s, t), 0);
    }

    #[test]
    fn test_s_shorter_than_t_with_some_matching_characters() {
        let s = String::from("ace");
        let t = String::from("abcdef");
        assert_eq!(Solution::append_characters(s, t), 3);
    }

    #[test]
    fn test_s_longer_than_t_with_t_scattered_within_s() {
        let s = String::from("abacadaeaf");
        let t = String::from("aaa");
        assert_eq!(Solution::append_characters(s, t), 0);
    }

    #[test]
    fn test_all_characters_in_t_are_in_s_but_not_in_order() {
        let s = String::from("abacadaeaf");
        let t = String::from("fa");
        assert_eq!(Solution::append_characters(s, t), 1);
    }

    #[test]
    fn test_partial_match_at_beginning_of_s() {
        let s = String::from("abc");
        let t = String::from("abf");
        assert_eq!(Solution::append_characters(s, t), 1);
    }

    #[test]
    fn test_partial_match_at_end_of_s() {
        let s = String::from("abcdef");
        let t = String::from("defg");
        assert_eq!(Solution::append_characters(s, t), 1);
    }
}


pub struct Solution;

impl Solution {
    pub fn append_characters(s: String, t: String) -> i32 {
        let mut point1: usize = 0;
        let mut point2: usize = 0;

        while point1 < s.len() && point2 < t.len() {
            if &s[point1..point1 + 1] == &t[point2..point2 + 1] {
                point1 += 1;
                point2 += 1;
            } else {
                point1 += 1;
            }
        }

        (t.len() - point2) as i32
    }
}